### DISCUSSION TITLE
Can NextJS, Next Auth be used with Apollo Client and GraphQl?

### DISCUSSION THREAD

**Discussion by akashshyamdev** [nextauthjs/next-auth#2915] [Help] [answered]
Can NextJS, Next Auth be used with Apollo Client and GraphQl, Mongoose? If so, could you point me to some resources?

**Comment by balazsorban44** [comment, marked as answer]
yes, absolutely! a quick Google returned several results

**Reply by akashshyamdev** [reply]
@balazsorban44 maybe we should add something about that in the docs? I did not find anything inside the docs.

**Reply by balazsorban44** [reply]
Sounds good! Feel free to create/link to a tutorial under https://next-auth.js.org/tutorials :slightly_smiling_face:

**Reply by akashshyamdev** [reply]
@balazsorban44 I can't really see a lot of info about using NextJS, next auth, graphql and *MONGOOSE*

**Reply by balazsorban44** [reply]
that stack is very specific, but you can puzzle together something from search results.

basically you want to do:

1. get the access token with any of our providers
2. return that access token in the session https://next-auth.js.org/configuration/callbacks
3. attach session to Apollo
4. make requests to your protected graphql api endpoint


As for mongoose, if it's to communicate with graphql, it does not concern next-auth.

If you want to save user data in a database, check out our adapters. we have one specific for mongodb, good resource to create your custom one specific for mongoose https://next-auth.js.org/adapters/mongodb

(note, this is for v4 of next-auth. v3 has been a bit different and used typeorm for mongodb support under the hood


https://next-auth.js.org/tutorials/creating-a-database-adapter

https://next-auth.js.org/v3/tutorials/creating-a-database-adapter

hope this helps!

**Reply by akashshyamdev** [reply]
Thanks a lot mate! btw, the adapter docs say that v4 is in beta https://next-auth.js.org/adapters/mongodb. But I don't see it in other places, is this an old warning you guys forgot to remove or is it still in beta?

**Reply by balazsorban44** [reply]
still beta. As said, the mongodb adapter did not exist in v3. it used the typeorm adapter, but it was easier to split it up and create a mongodb specific one.

**Reply by akashshyamdev** [reply]
@balazsorban44 but some places in the docs say that v3 support is dropped. Aren't we supposed to drop support after the beta is official? Also, most mongoose & next-auth tutorials/blog posts use the typeorm(v3) instead of v4.

**Reply by akashshyamdev** [reply]
As you said graphql is not related to nextjs, let's leave that for now. I'm trying to do the following:

1. Use OAuth providers like Google, Facebook, Twitter and Github
2. After the access token stuff is done, then create a user using mongoose inside the DB
3. Create a JWT based on user `_id`.
4. On the frontend, based on the JWT, implement signup and protected routes

I haven't worked with NextJS and I've used JWT based auth in react and nodeJS rest apis. Could you help me with this?

**Reply by balazsorban44** [reply]
I would start reading up on Next.js then and learning about JWTs. Unfortunately, I don't have the capacity to teach about those.

Some useful links: 
- https://nextjs.org/docs/getting-started
- https://next-auth.js.org/faq#json-web-tokens
- https://next-auth.js.org/tutorials
